### PR TITLE
simple_uart_describe

### PR DESCRIPTION
**Pull request by akaeba**
Add win32 implementation, return key/value string pair list. Linux implementation is WIP.

### CODE REVIEW

**Comment by AndreRenaud** [comment]
I sort of feel that the double indirect list of strings is an awkward API for people to use. It makes the chances of memory leaks quite high - that was the intention behind using `simple_uart_describe(const char *uart, char *description, size_t max_desc_len)` - it's pretty straight forward.

I do agree that the downside to the simple API is consistency between OSs, but since we're only talking about 2/3 options, it feels like it's ok to just manually manage that?

**Comment by akaeba** [comment]
Okay then do it as sind mole string, this reduces at least the code in simple_uart. For windows case I would for start only append the hardware id. The rest on request. For linux I would simply output the `udevadm` result. 

I can adjust for win and Linux. For Mac I have no clue.

**Comment by akaeba** [comment]
Hi Andre, please take a look. From my point of view is the WIndows/Linux part ready for merge. For MacOS we need volunteers ;)
BR,
Andreas

File: simple_uart.c [hunk 1]
```diff
@@ -518,6 +521,146 @@ ssize_t simple_uart_list(char ***namesp)
...
+            }
+        } else {
+            r = -GetLastError();
+            if (r > INT_MAX) r = INT_MAX;
+            return (int)r;
+        }
+        ++i;
+    }
+    /* no match by name */
+    if ( 0 == intNameMatch )
```
Comment (akaeba) [inline simple_uart.c]: #31

File: simple_uart.c [hunk 2]
```diff
@@ -518,6 +521,146 @@ ssize_t simple_uart_list(char ***namesp)
...
+        return 0;
```
Comment (akaeba) [inline simple_uart.c:568]: #31 see here

File: simple_uart.c [hunk 3]
```diff
@@ -518,6 +521,146 @@ ssize_t simple_uart_list(char ***namesp)
...
+        }
+        if ( 0 != intMatchDev ) {
+            break;
+        }
+    }
+    pclose(fH2);
+    pclose(fH1);
+    /* device found? */
+    if ( 0 == intMatchDev ) {
+        return -1;
```
Comment (akaeba) [inline simple_uart.c:640]: #31 and here

**Comment by akaeba** [comment]
From my point of view is the PR now ready.

**Comment by AndreRenaud** [comment]
Sorry, I've been busy and haven't had a chance to look at this. I'll try and check it out in the next few days.

**Comment by akaeba** [comment]
No worry, thanks :). Please don't feel pushed by me

File: simple_uart.c [hunk 4]
```diff
@@ -518,6 +521,141 @@ ssize_t simple_uart_list(char ***namesp)
...
+    if (SetupDiGetDeviceRegistryPropertyA(deviceInfoSet, &deviceInfoData, SPDRP_HARDWAREID, NULL, (PBYTE)chrBuf, sizeof(chrBuf), NULL)) {
+        if ( !((strlen(chrBuf) + strlen(description) + 3) < max_desc_len) ) // check for enough memory, +3 '\n'
+            return -1;
+        strncpy(description+strlen(description), chrBuf, strlen(chrBuf)+1); // copy string
+        description[strlen(description)+1] = '\0';
+        description[strlen(description)] = '\n';    // every new property gets a new line
+    }
+    // Append next required property here
+// Linux Implementation
+#elif defined(__linux__)
```
Comment (AndreRenaud) [inline simple_uart.c:581]: I'm not super keen on shelling out like this - it make assumptions about what's available on the OS, what shell the user is running etc...

I think similar functionality can be achieved by reading the `manufacturer`, `product`, `idProduct`, `idVendor` files that would be available in `/sys/class/tty/ttyUSB0/device/../..` (some realpath trickery is required here to turn the tty device name into the USB device path, as these are distinct`.

I bodged together this standalone program to demonstrate:
```c
#include <limits.h>
#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>
#include <libgen.h>
#include <string.h>

static bool readfile(const char *dir, const char *filename, char *result, size_t max_result_len) {
	FILE *fp;
	size_t len = 0;
	char path[PATH_MAX];
	snprintf(path, sizeof(path), "%s/%s", dir, filename);
	result[0] = '\0';
	if ((fp = fopen(path, "rb")) != NULL) {
		len = fread(result, 1, max_result_len - 1, fp);
		if (len) {
			result[len - 1] = '\0';
		}
		fclose(fp);
	}
	return len != 0;
}


int serial_uart_describe(const char *uart, char *description, size_t max_len)
{
	char devPath[PATH_MAX];
	char basePath[PATH_MAX];
	size_t pos = 0;

	description[0] = '\0';
	snprintf(basePath, sizeof(basePath), "/sys/class/tty/%s/device/../..", basename((char *)uart));
	if (realpath(basePath, devPath) != NULL) {
		char contents[256];

		if (readfile(basePath, "manufacturer", contents, sizeof(contents))) {
			pos += snprintf(&description[pos], max_len - pos, "%smanufacturer='%s'", pos != 0 ? "," : "", contents);
		}
		if (readfile(basePath, "product", contents, sizeof(contents))) {
			pos += snprintf(&description[pos], max_len - pos, "%sproduct='%s'", pos != 0 ? "," : "", contents);
		}
		if (readfile(basePath, "idProduct", contents, sizeof(contents))) {
			pos += snprintf(&description[pos], max_len - pos, "%sPID=%s", pos != 0 ? "," : "", contents);
		}
		if (readfile(basePath, "idVendor", contents, sizeof(contents))) {
			pos += snprintf(&description[pos], max_len - pos, "%sVID=%s", pos != 0 ? "," : "", contents);
		}
		if (readfile(basePath, "serial", contents, sizeof(contents))) {
			pos += snprintf(&description[pos], max_len - pos, "%sserial=%s", pos != 0 ? "," : "", contents);
		}
	}
	return 0;
}

int main(int argc, char **argv) {
	char descr[1024];
	const char *dev = "/dev/ttyUSB0";

	if (argc > 1) {
		dev = argv[1];
	}


	serial_uart_describe(dev, descr, sizeof(descr));

	printf("description: %s\n", descr);

	return 0;
}
```

This results in output like:
```sh
$ gcc -o describe describe.c  -g -Wall -O2 && ./describe /dev/ttyUSB0
description: manufacturer='Prolific Technology Inc.',product='USB-Serial Controller',PID=2303,VID=067b
```

Any thoughts on this?

Comment (andkae) [reply simple_uart.c:581]: Also suitable! And you are definitly right thats the more generic linux way. Im for this implementation. Should i implemement or do you like to adjust my branch?

BR,
Andreas

Comment (akaeba) [reply simple_uart.c:581]: I can adjust my PR on weekend according your proposal. Stay tuned.

Comment (AndreRenaud) [reply simple_uart.c:581]: Thanks for that - sorry, I haven't had time to come back to this. I don't really have any comments on the Windows aspect, I'm not familiar enough with their APIs. If you could dump some example descriptions into the PR, that'd be great.

For macos, I'll worry about that later :)

**Comment by akaeba** [comment]
Hi Andre, i applied your proposal.

```snprintf(basePath, sizeof(basePath), "/sys/class/tty/%s/device/../..", basename((char *)uart));``` On my machine i recieved for ```/dev/ttyACM0``` the result ```manufacturer='Linux 5.11.0-37-generic ohci_hcd',product='OHCI PCI host controller',PID=0001,VID=1d6b,serial=0000:00:06.0```. Therefore i adjusted your path building.

**Comment by akaeba** [comment]
Now it's working with two adapters:
* manufacturer='Prolific Technology Inc.',product='USB-Serial Controller',PID='2303',VID='067b'
* manufacturer='Devantech Ltd.',product='USB-ISS.',PID='ffee',VID='04d8',serial='00060147'
*

**Comment by akaeba** [comment]
and under windows:
* hwid='USB\VID_04D8&PID_FFEE&REV_0100'

**Comment by akaeba** [comment]
After adjustment looks the windows string ```PID=FFEE,VID=04D8```. I check at the moment the WIN API for gathering information comparable to product and manufacturer under linux. But it seems a little bit difficult.

**Comment by AndreRenaud** [comment]
Is this good to merge now?

**Comment by akaeba** [comment]
Yes bring to main, and if i find something related to Manufacturer i create a new PR.

**Comment by AndreRenaud** [comment]
Thanks for all your work on this

**Comment by akaeba** [comment]
Thanks for merge :D

**Pull request merged**